core/sgf/parser.go

```diff
@@ -0,0 +1,399 @@
+//
+// Transitions:
+//
+//     between => propertyState        ex: AW
+//     between => propDataState        ex: AW[ab][
+//     between => between, add branch  ex: B[ab](
+//     between => between, pop branch  ex: B[ab](;W[ac])
+//     between => between, add node    ex: B[ab];
+func handleBetween(stateData *stateData, pbuf *propBuffer) error {
+	if unicode.IsSpace(stateData.curchar) {
+		// We can safely ignore whitespace here.
+		return nil
+	} else if unicode.IsUpper(stateData.curchar) {
+		// AW[aw][bw]
+		// ^
+		if err := pbuf.flush(stateData.curnode); err != nil {
+			return stateData.parseError(err.Error())
+		}
+		stateData.addToBuf(stateData.curchar)
+		stateData.curstate = propertyState
+		return nil
+	} else if stateData.curchar == lbrace {
+		// AW[aw][bw]
+		//   ^   ^
+		stateData.curstate = propDataState
+		return nil
+	} else if stateData.curchar == lparen {
+		// AW[aw][bw] (;B[ab]
+		//            ^
+		pbuf.flush(stateData.curnode)
+		if err := pbuf.flush(stateData.curnode); err != nil {
+			return stateData.parseError(err.Error())
+		}
+		stateData.addBranch(stateData.curnode)
+		return nil
+	} else if stateData.curchar == scolon {
+		// AW[aw][bw] (;B[ab];W[ac])
+		//             ^     ^
+		pbuf.flush(stateData.curnode)
+		if err := pbuf.flush(stateData.curnode); err != nil {
+			return stateData.parseError(err.Error())
+		}
+		cn := stateData.curnode
+		stateData.curnode = game.NewNode()
+		cn.AddChild(stateData.curnode)
+		stateData.curnode.Parent = cn
+		return nil
+	} else if stateData.curchar == rparen {
+		// AW[aw][bw] (;B[ab])
+		//                   ^
+		pbuf.flush(stateData.curnode)
+		if err := pbuf.flush(stateData.curnode); err != nil {
+			return stateData.parseError(err.Error())
+		}
+		cn, err := stateData.popBranch()
+		if err != nil {
+			return err
+		}
+		stateData.curnode = cn
+		return nil
+	}
+	return stateData.parseError("unexpected character between")
+}
+
+// handleProperty is a simple state for handling parsing property
+// keys (AW, C, etc)
+//
+// Transitions:
+//
+//     property => propData
+func handleProperty(stateData *stateData, pbuf *propBuffer) error {
+	if unicode.IsUpper(stateData.curchar) {
+		// AW[aw][bw]
+		//  ^
+		stateData.addToBuf(stateData.curchar)
+		return nil
+	} else if stateData.curchar == lbrace {
+		// AW[aw][bw]
+		//   ^
+		prop := stateData.flushBuf()
+		pbuf.prop = prop
+		stateData.curstate = propDataState
+		return nil
+	}
+	return stateData.parseError("unexpected character during property parsing")
+}
+
+// handlePropData is a state for handling parsing property
+// data
+//
+// Transitions:
+//
+//     propData => propData
+//     propData => between
+func handlePropData(stateData *stateData, pbuf *propBuffer) error {
+	if stateData.curchar == rbrace &&
+		// C[foo 1[k\] bar]
+		//           ^
+		// There was a backslash used for escaping a r-brace.
+		stateData.holdChar == backslash {
+		stateData.holdChar = rune(0)
```

Comment:
Is the formatting off in these if-statements? Part of the if statement conditional is in line with the if branch. That seems wrong.

Reply:
It's technically correct, but confusing. Fixed